### ISSUE TITLE
Dense Layer Input Property

### ISSUE THREAD

**Issue by Cliff-Foreman** [losttech/Gradient#26]
Hi everyone, loving Gradient so far, but I ran into a problem. I was making just a simple sequential model

```csharp
var model = new Sequential();
var hidden1 = new Dense(5, activation: "relu");
var hidden2 = new Dense(1, activation: "sigmoid");
model.add(hidden1);
model.add(hidden2);
```

But, I noticed that the Dense function did not have input_shape or input_dim property. So, I was wondering how I could pass input into the sequential model without the property.

Thanks in advanced

**Comment by lostmsu** [comment]
@Cliff-Foreman you should be able to pass them in a dictionary to `kwargs`: [example in FashionMNIST](https://github.com/losttech/Gradient-Samples/blob/ad6dada004da95852b38ed44efdaf37f75dca991/FashionMnistClassification/FashionMnistClassification.cs#L30).

**Comment by Cliff-Foreman** [comment]
So, I'm trying to create a XOR solver, and want to pass an array with two elements (0,0 for instance). So, should I just create a python dictionary of that list and pass it in: 

```csharp
var hidden1 = new Dense(5, activation: "relu",
                        kwargs: new PythonList<int,object> { ["input_shape"] = (2) });
```

Because that gives an error. Sorry, I'm very new to this.

**Comment by lostmsu** [comment]
You need to attach the error message to get help quicker.

In this case you made a mistake: it must be `PythonDict<string, object>`. Or you could use regular `Dictionary<string, object>`. This nearly always will be a type for `kwargs` because they are essentially name=value pairs, where name is a `string` and value can be of any type (in C# `object` is a base type for everything else, and so can hold any value).

**Comment by Cliff-Foreman** [comment]
Yeah, sorry for not attaching the error. Here is my error: 

Unhandled exception. System.TypeInitializationException: The type initializer for 'numpy.np' threw an exception.
 ---> System.MissingMethodException: Method not found: 'Void Python.Runtime.Runtime.set_PythonDLL(System.String)'.
   at Gradient.GradientEngine.PreloadPythonEmbeddingLibrary(PythonEnvironment pythonEnvironment)
   at Gradient.GradientEngine.Initialize()
   at Gradient.GradientEngine.EnsureInitialized()
   at Gradient.GradientSetup.EnsureInitialized()
   at numpy.np..cctor()
   --- End of inner exception stack trace ---
   at numpy.np.array(Object object, Object dtype, Boolean copy, String order, Boolean subok, Int32 ndmin)
   at First.Program.Main(String[] args) in /Users/Cliff/Projects/Grad/Grad/Program.cs:line 18

and here is my code: 
```csharp
using System;
using tensorflow;
using Gradient;
using tensorflow.keras;
using tensorflow.keras.layers;
using numpy;
using Gradient.BuiltIns;

namespace First
{
    class Program
    {
        static void Main(string[] args)
        {
            GradientLog.OutputWriter = Console.Out;
            GradientEngine.UseEnvironmentFromVariable();

            ndarray x = np.array(new int[,] { { 0, 0 }, { 0, 1 }, { 1, 0 }, { 1, 1 } });
            ndarray y = np.array(new int[] { 0, 1, 1, 0 });

            var model = new Sequential();
            var hidden1 = new Dense(5, activation: "relu", kwargs: new PythonDict<string, object> { ["input_shape"] = (2) });
            var hidden2 = new Dense(1, activation: "sigmoid");

            model.add(hidden1);
            model.add(hidden2);

            model.compile(optimizer: "adam", loss: "binary_crossentropy");
            model.fit(x, y, epochs: 100);

            ndarray test = np.array(new int[] { 0, 0 });
            model.predict(test);
            var text = model.predict(test);
            Console.WriteLine(test);
        }
    }
}
```
Thanks in advance

**Comment by lostmsu** [comment]
The error looks like you somehow ended up using `Python.Runtime.dll`, that did not come with Gradient. E.g. you have another package referenced in your project, that brings its own older version of `Python.Runtime.dll`.

**Comment by lostmsu** [comment]
BTW, in C# `(2)` is the same as just a number `2`, not a tuple with one element. I recommend doing `new TensorShape(2)` instead for code clarity. Or `ValueTuple.Create(2)` if you want a tuple specifically. `(2,2)` gives back a tuple though.

**Comment by Cliff-Foreman** [comment]
Ah, thanks for the advice on the tensor shape. I'm using a macbook, so I'm wondering how exactly I can load the python runtime dll. Do I need to add the pythonnet package on Visual Studio?

**Comment by lostmsu** [comment]
@Cliff-Foreman normally, when you reference Gradient package in Visual Studio, it should automatically download corresponding `Python.Runtime.dll` from [Python.Runtime.UnofficialNetStandard](https://www.nuget.org/packages/Python.Runtime.UnofficialNetStandard/) package, because it is listed in dependencies. The problem you have though, is that you also referenced some other package, that also has `Python.Runtime.dll` in it, but of an older version, and that old version gets loaded instead. E.g. there's a package conflict with some other package.

Remove unused `<PackageReference>` elements from your .csproj file. You can share your .csproj file here, so I could investigate if that does not help.

**Comment by Cliff-Foreman** [comment]
Yeah, here is my .csproj file

<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Gradient" Version="0.15.7.2" />
  </ItemGroup>
</Project>

**Comment by Cliff-Foreman** [comment]
I added the python reference to the csproj, and now I get this error: 

Unhandled exception. Python.Runtime.PythonException: TypeError : 'int' object is not iterable
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tensorflow_core/python/keras/layers/core.py", line 988, in __init__
    activity_regularizer=regularizers.get(activity_regularizer), **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tensorflow_core/python/training/tracking/base.py", line 457, in _method_wrapper
    result = method(self, *args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/base_layer.py", line 363, in __init__
    batch_input_shape = (batch_size,) + tuple(kwargs['input_shape'])

**Comment by lostmsu** [comment]
@Cliff-Foreman yeah, that one is due to the `(2)` thing I mentioned before.
Also, you might want to change your input and output `np.array`s from `int` to `float`.

**Comment by Cliff-Foreman** [comment]
Hi @lostmsu yeah, that got the error to go away. However, I'm having a problem with my numpy now. My numpy array is: 

```csharp
ndarray x = np.array(new float[,] { { 0, 0 }, { 0, 1 }, { 1, 0 }, { 1, 1 } });
```

But, when I Console.Write(x.shape), the shape is 8. I was wondering how this can be changed to a 2d array. Btw, I did change it to float. Thanks

**Comment by lostmsu** [comment]
That seems like a bug at least because the behavior is unexpected. You can either pass a jagged array (e.g. `float[][]` instead of `float[,]`), or call `reshape`.

P.S. Use triple backticks to paste code blocks. E.g.

```
```csharp // <-- language name here
Console.WriteLine(..);
return 10;
```/do not add this/
```

**Comment by Cliff-Foreman** [comment]
I tried using float[][], but when my line is: 

```csharp //
ndarray x = np.array(new float[][] { { 0, 0 }, { 0, 1 }, { 1, 0 }, { 1, 1 } });
```

But, Visual Studio flags this as an error.

Btw, the triple backtick is really cool

**Comment by lostmsu** [comment]
@Cliff-Foreman a bit uglier, but works (`0f` makes it `float`):
```csharp
np.array(new [] { new[]{ 0f, 0 }, new []{ 0f, 1 }, new []{ 1f, 0 }, new []{ 1f, 1 } });
```
I opened a [bug for `[,]` arrays](https://github.com/losttech/Gradient/issues/27).

**Comment by Cliff-Foreman** [comment]
Yeah, that finally got it working. Thank you so much for your help. You were truly helpful, and I am excited with using Gradient. Thanks.

**Comment by lostmsu** [comment]
@Cliff-Foreman you're welcome.
